draft-ralston-mimi-linearized-matrix.md

```diff
@@ -687,17 +687,215 @@ The hub server MUST enforce the room version's algorithms upon the room. Partici
 
-Each room has a single MLS Group associated with it, starting with the `m.room.create` event.
+Rooms additionally track membership at a per-user level while MLS tracks group membership at a
+per-device level. With this consideration, commits to the MLS Group MUST use `PublicMessage`, giving
+the hub server an ability to inspect MLS group membership changes for illegal joins and leaves.
 
-**TODO**: Details on how key material is stored/shared within the room.
+Encryption can only be enabled at the time the room is created. This prevents the room having encryption
+disabled or downgraded without an entirely new room being created. The exact ciphersuite and other
+algorithmic details are contained in the `content` for the `m.room.create` event ({{int-ev-create}}):
 
-**TODO**: What does `m.room.encrypted` (user message) look like here?
+~~~ json
+{
+   "encryption": {
+      "algorithm": "m.mls.v1.dhkemx25519-aes128gcm-sha256-ed25519"
+   }
+}
+~~~
+
+`algorithm` denotes which specific algorithm clients MUST use for sending and receiving encrypted
+events in the room. If a received event is encrypted using a different algorith, it MUST be treated
+as undecryptable (even if the client has sufficient key information to decrypt it).
+
+`m.mls.v1.` as a prefix describes the behaviour for encrypted clients, with the remainder of the
+algorithm string covering the exact ciphersuite. This document uses the same mandatory ciphersuite
+as MLS: `MLS_128_DHKEMX25519_AES128GCM_SHA256_Ed25519`. Thus, this is encoded as
+`m.mls.v1.dhkemx25519-aes128gcm-sha256-ed25519`. Other ciphersuites can be represented similarly,
+though are considered to be entirely new encryption algorithms for the purposes of this document.
+
+Custom or non-standard encryption algorithms are possible with this approach, however out of scope
+for MIMI. If such an algorithm is used, it SHOULD be prefixed using reverse domain name notation.
+For example, `org.example.my-encryption`.
+
+Mentioned in the introduction ({{int-intro}}), this document does not explore the details for what is
+needed to interconnect Linearized Matrix and Matrix's existing room model. However, for interconnection
+to be successful, extensions to MLS are needed to support decentralization. One possible extension
+is "Decentralised MLS" {{DMLS}}.
+
+## Device Credentials
+
+Under Section 5.3 of {{!I-D.ietf-mls-protocol}}, each MLS group member (a device, {{int-device-id}})
+has a "credential" or signing key associated with it. These are published to each client's local server
+and available over federation ({{int-transport-mls}}).
+
+This document relies upon out-of-band verification and therefore uses basic credentials. The format
+for the credential is:
+
+~~~
+struct {
+    opaque identity<V>;
+    SignatureScheme signature_scheme;
+    opaque signature_key<V>;
+} BasicCredential;
+~~~
+
+**TODO**: Verify `BasicCredential` format.
```

Comment:
Reviewer: please verify

Reply:
My thinking was that the BasicCredential would contain separate user ID and device ID fields.  Though I don't know if there's a better way to it.

Reply:
I borrowed this from a much earlier draft of the MLS protocol, so not sure it makes much sense either. I imagine `identity` is the only bit which needs to be split apart?

Reply:
Yeah, probably just need to split `identity` into two fields.  Might also be able to drop the `SignatureScheme` since we don't have that enum any more in MLS.